repl/commands.go

```diff
@@ -16,117 +16,289 @@
-			return
+		for _, p := range ctx.GetParams() {
+			if p.GetT() == nil {
+				c.reportIssue(errors.New("missing type in function param declaration"))
+				continue
+			}
+			if p.GetPid() == nil {
+				c.reportIssue(errors.New("missing identifier in function param declaration"))
+			}
+			ty, err := ParseType(p.GetT().GetText())
+			if err != nil {
+				c.reportIssue(err)
+			}
+			cmd.params = append(cmd.params, letFunctionParam{
+				identifier: p.GetPid().GetText(),
+				typeHint:   ty,
+			})
+
 		}
-		if i := subexpByName(delRe, m, "ident"); i != nil {
-			args = append(args, *i)
+		cmd.resultType = ty
+	case *delCmd:
+		if ctx.GetId() == nil {
+			c.reportIssue(errors.New("missing identifier in delete"))
 		}
-	case "declare":
-		m := declRe.FindStringSubmatch(line)
-		if m == nil {
-			err = PromptError{msg: "invalid declare statement", usage: &declareUsage}
+		cmd.identifier = ctx.GetId().GetText()
+	default:
+		c.reportIssue(errors.New("unexepected function declaration"))
+	}
+}
+
+func (c *commandParseListener) ExitQualId(ctx *parser.QualIdContext) {
+	if ctx.GetRid() == nil {
+		c.reportIssue(errors.New("missing root identifier"))
+	}
+}
+
+// extract source text from a parse rule match.
+// preserves original whitespace if possible
+func extractSourceText(ctx antlr.ParserRuleContext) string {
+	if ctx.GetStart() == nil || ctx.GetStop() == nil ||
+		ctx.GetStart().GetStart() < 0 || ctx.GetStop().GetStop() < 0 {
+		// fallback to the normalized parse
+		return ctx.GetText()
+	}
+	s, e := ctx.GetStart().GetStart(), ctx.GetStop().GetStop()
+	return ctx.GetStart().GetInputStream().GetText(s, e)
+}
+
+func (c *commandParseListener) ExitVarDecl(ctx *parser.VarDeclContext) {
+	switch cmd := c.cmd.(type) {
+	case *letVarCmd:
+		if ctx.GetId() == nil {
+			c.reportIssue(errors.New("no identifier in variable declaration"))
 			return
 		}
-		if a := subexpByName(declRe, m, "ident"); a != nil {
-			args = append(args, *a)
+		cmd.identifier = ctx.GetId().GetText()
+		if ctx.GetT() != nil {
+			ty, err := ParseType(ctx.GetT().GetText())
+			if err != nil {
+				c.reportIssue(err)
+			}
+			cmd.typeHint = ty
 		}
-		if t := subexpByName(declRe, m, "type_hint"); t != nil {
-			args = append(args, *t)
+	case *delCmd:
+		if ctx.GetId() == nil {
+			c.reportIssue(errors.New("missing identifier in delete"))
 		}
-	case "eval":
-		expr = line
-	case "exit":
-	case "null":
+		cmd.identifier = ctx.GetId().GetText()
 	default:
-		err = PromptError{msg: "Undefined command: " + cmd}
+		c.reportIssue(errors.New("unexpected var declaration"))
 	}
-	return
+}
+
+func (c *commandParseListener) ExitExpr(ctx *parser.ExprContext) {
+	expr := extractSourceText(ctx)
+	switch cmd := c.cmd.(type) {
+	case *evalCmd:
+		cmd.expr = expr
+	case *letFnCmd:
+		cmd.src = expr
+	case *letVarCmd:
+		cmd.src = expr
+	default:
+		c.reportIssue(errors.New("unexpected CEL expression"))
+	}
+}
+
+func Parse(line string) (Cmd, error) {
```

Comment:
Don't forget to comment this method. You may want to group the other ANTLR method after the `Parse` with a general heading that indicates these are ANTLR listener methods.